File: 2024/day5/day5.R
library(tidyverse)

# Load and parse the input file
input <- readLines("input_day5.txt")

# Split into rules and updates based on the blank line
split_index <- which(input == "")[1]  # Find the blank line separating rules and updates
rules_raw <- input[1:(split_index - 1)]  # Rules section
updates_raw <- input[(split_index + 1):length(input)]  # Updates section

# Parse rules into a tibble
rules <- rules_raw |>
  as_tibble() |>
  separate(value, into = c("page_before", "page_after"), sep = "\\|") |>
  mutate(
    page_before = as.numeric(page_before),
    page_after = as.numeric(page_after)
  ) |>
  group_by(page_before) |>
  summarize(page_after = list(page_after), .groups = "drop")

# Parse updates into a list of numeric vectors
updates <- updates_raw |>
  map(~ as.numeric(str_split(.x, ",")[[1]]))

# Function to check if an update is in the correct order
is_in_correct_order <- function(pages) {
  l <- length(pages)
  for (i in 1:(l - 1)) {
    possible_pages <- rules |>
      filter(page_before == pages[i]) |>
      pull(page_after) |>
      unlist()
    remaining_pages <- pages[(i + 1):l]
    
    if (!all(map_lgl(remaining_pages, ~ .x %in% possible_pages))) {
      return(FALSE)
    }
  }
  return(TRUE)
}

# Function to extract the middle page of a correctly ordered update
get_middle_page <- function(pages) {
  pages[ceiling(length(pages) / 2)]
}

# Part 1: Sum of middle pages of updates already in the correct order
part1_result <- updates |>
  keep(is_in_correct_order) |>
  map_dbl(get_middle_page) |>
  sum()

print(part1_result)

# Function to reorder an update to make it correct
reorder_update <- function(pages) {
  if (is_in_correct_order(pages)) {
    return(pages)
  }
  
  l <- length(pages)
  while (!is_in_correct_order(pages)) {
    for (i in 1:(l - 1)) {
      possible_pages <- rules |>
        filter(page_before == pages[i]) |>
        pull(page_after) |>
        unlist()
      
      for (j in (i + 1):l) {
        if (!pages[j] %in% possible_pages) {
          pages <- c(pages[j], pages[-j])  # Move the out-of-place page to the front
          break
        }
      }
    }
  }
  return(pages)
}

# Part 2: Sum of middle pages of reordered updates
part2_result <- updates |>
  discard(is_in_correct_order) |>
  map(reorder_update) |>
  map_dbl(get_middle_page) |>
  sum()

print(part2_result)

File: 2024/day05/Python/day5.py
from collections import defaultdict

# Load
with open("input_day5.txt") as file:
    input_data = file.read().strip()

# Split rules and updates sections
rules_section, updates_section = input_data.split("\n\n")

# Parse rules
page_rules = defaultdict(set)
for rule_line in rules_section.split("\n"):
    page_before, page_after = map(int, rule_line.split("|"))
    page_rules[page_before].add(page_after)

# Parse updates
updates_list = [list(map(int, update_line.split(","))) for update_line in updates_section.split("\n")]


def is_in_correct_order(update):
    remaining_pages = set(update)  # Track remaining pages to check
    for current_page in update:
        remaining_pages.remove(current_page)
        if not (remaining_pages <= page_rules[current_page]):
            return False
    return True


# Part 1
correctly_ordered_updates = [
    update[len(update) // 2] for update in updates_list if is_in_correct_order(update)
]
print(sum(correctly_ordered_updates))


def reorder_update(update):
    index = 0
    while index < len(update) - 1:
        current_page = update[index]
        remaining_pages = set(update[index + 1:])
        if not (remaining_pages <= page_rules[current_page]):
            # Move the out-of-place page to the end
            update.append(update.pop(index))
            # Restart checking from the beginning
            index = 0
        else:
            index += 1
    return update


# Part 2
incorrect_updates = [
    update for update in updates_list if not is_in_correct_order(update)
]
reordered_updates = [reorder_update(update) for update in incorrect_updates]
reordered_middle_pages = [
    update[len(update) // 2] for update in reordered_updates
]
print(sum(reordered_middle_pages))
